cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,492 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
+   * `maxSteps` bound, cuts the list of nodes that are to be added to the queue
+   * in order not to add too many of them.
+   *
+   * Assumes that all nodes in the queue are being visited.
+   */
+  def limitAddedToQueue(nodes: Seq[Int]): Seq[Int] = {
+    maxSteps match {
+      case None => nodes
+      case Some(max) => if (max - numEnqueuedEver > Int.MaxValue) {
+        nodes
+      } else {
+        nodes.take((max - numEnqueuedEver).toInt)
+      }
     }
-    if (!seen) distanceTracker.recordInfo(id, dist)
   }
 
-  protected def storeForVisit(id: Int, info: T)
+  /**
+   * Returns nodes ids that should be added to the queue after visiting `node`.
+   */
+  def chooseNodesToEnqueue(node: Node): Seq[Int] = {
+    val currDepth = depth(node.id).get
 
-  protected def seenBefore(id: Int) = distance(id).isDefined
+    if (depthLimit.isLimitReached(currDepth) ||
+      degreeLimit.isLimitReached(node.neighborCount(dir))) {
+      Seq()
+    } else {
+      limitAddedToQueue(node.neighborIds(dir).filter(n => shouldBeEnqueued(coloring.get(n))))
+    }
+  }
+
+  /**
+   * Enqueues `nodes` added after `from` node is visited (or None
+   * when equeueing initial nodes).
+   */
+  protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    numEnqueuedEver += nodes.size
+    nodes.foreach {
+      node => coloring += (node, NodeColor.Enqueued)
+    }
+    if (prevNbrCounter.isDefined && from.isDefined) {
+      nodes.foreach {
+        node => prevNbrCounter.get.recordPreviousNeighbor(node, from.get)
+      }
+    }
+    nodePriority match {
+      case GraphTraverserNodePriority.LIFO =>
+        nodes.reverse.foreach (node => queue.enqueueFirst(node))
+      case GraphTraverserNodePriority.FIFO =>
+        nodes.foreach (node => queue.enqueue(node))
+    }
+  }
+
+  /**
+   * Performs action needed when visiting a node.
+   */
+  protected def processNode(node : Node): Node = {
+    visitNode(node)
+    coloring += (node.id, NodeColor.Visited)
+    enqueue(chooseNodesToEnqueue(node), Some(node.id))
+    node
+  }
+
+  /**
+   * Finds in the queue the node that will be visited next.
+   */
+  protected def findNextNodeToVisit(): Option[Int] = {
+    if (queue.isEmpty)
+      None
+    else
+      Some(queue.firstInt())
+  }
+
+  def next() = {
+    findNextNodeToVisit() match {
+      case Some(nextId) =>
+        if (shouldBeDequeuedBeforeProcessing) queue.dequeueInt()
+        processNode(getExistingNodeById(graph, nextId))
+      case None =>
+        Iterator.empty.next()
+    }
+  }
 
-  def distance(id: Int) = distanceTracker.infoOfNode(id)
+  def hasNext = findNextNodeToVisit().isDefined
 }
 
 /**
- * Traverses in breadth-first order. This implies that first all the neighbors of
- * {@code homeNodeIds} are visited, then their neighbors, etc.
+ * A traverser that keeps track of first depth of visiting a given node.
+ */
+trait DepthTracker extends QueueBasedTraverser {
```

Comment:
May be a small point but should all trackers be of type mixin traits? i.e., trait DepthTracker { this: QueueBasedTraverser => ... } rather than extending QueueBasedTraverser.
http://baddotrobot.com/blog/2014/09/22/scala-mixins/